HIP/HIP-668.md

```diff
@@ -0,0 +1,2349 @@
+an `after` or `before` arguments to perform either ascending or descending order pagination, respectively. While
+we feel the edge wrapper is a bit unnecessary, following the specification as is provides for a good out of the box
+experience with most GraphQL clients.
+
+### Queries
+
+The current REST API has two broad top-level query patterns: list all entities/transactions or get a specific
+entity/transaction. Based upon metrics, the latter scenario is a lot more common since dApps
+will generally want to look up their specific account, topic, transaction, etc. Alternatively, they want to list all
+data specific to their entity like all transactions involving a certain account. Listing across accounts or other top
+level resources makes more sense for narrower use cases like explorers or analytics.
+
+With upcoming work to shard the database by its Hedera entity, queries that involve many entities will be less
+efficient to execute since they will involve fetching data from all database shards spread across many physical database
+instances. For these reasons, we will avoid creating list APIs that paginate across entities and focus our efforts on
+paginating within a specific entity. In the future, we could entertain adding such a feature, but it would be most
+likely constrained to the last `X` or top `N` type queries and not allow pagination to allow for caching.
+
+```graphql
+"The query root of the Hedera GraphQL API"
+type Query {
+  account(id: AccountIdInput!): Account
+  block(id: BlockIdInput!): Block
+  contract(id: ContractIdInput!): Contract
+  file(id: FileIdInput!): File
+  network: Network!
+  node(id: ID! @Pattern(regexp: "^[a-zA-Z0-9+/=]+$")): Node
+  schedule(id: ScheduleIdInput!): Schedule
+  token(id: TokenIdInput!): Token
+  topic(id: TopicIdInput!): Topic
+  transaction(id: TransactionInput!): Transaction
+}
+
+```
+
+### Directives
+
+Below are the custom directives that aid in validating input in a declarative manner. Providing validation directives
+makes it explicit in the schema the input requirements instead of requiring an out-of-band communication of validation
+criteria.
+
+```graphql
+"Validation directive that ensures the argument or input is a certain minimum value."
+directive @Min(value : Int! = 0, message : String = "graphql.validation.Min.message")
+on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION
+
+"Validation directive that ensures the argument or input value matches the given regular expression."
+directive @Pattern(regexp : String! =".*", message : String = "graphql.validation.Pattern.message")
+on ARGUMENT_DEFINITION | INPUT_FIELD_DEFINITION
+
+```
+
+### Scalars
+
+In addition to the built-in scalars, this API defines a few custom scalars.
+
+```graphql
+"An ISO 8601 compatible duration with support for nanoseconds granularity in the format P[n]Y[n]M[n]DT[n]H[n]M[n]S."
+scalar Duration
+
+"A 64-bit numeric type."
```

Comment:
Signed or unsigned?  I think it should be explicit.

Reply:
True, will add a note that it's signed. Built-in types like `Int` are signed and this one will be as well, especially since it's backed by Java which only has signed.